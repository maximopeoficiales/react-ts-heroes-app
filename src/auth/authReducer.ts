export enum Action {
    LOGIN = "login",
    LOGOUT = "logout",
    DEFAULT = "",
}
export type User =
    {
        name?: string;
        logged?: boolean;
    }

export type ReducerUser = { type: Action, payload?: User };

export const authReducer = (state: User, action: ReducerUser): User => {
    switch (action.type) {
        case Action.LOGIN:
            return { ...action.payload, logged: true }
        case Action.LOGOUT:
            return { logged: false };
        default:
            return state;
    }

}