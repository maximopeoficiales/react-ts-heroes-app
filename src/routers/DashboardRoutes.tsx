import React from "react";
import { Redirect, Route, Switch } from "react-router-dom";
import DCScreen from "../components/dc/DcScreen/DcScreen";
import HeroesScreen from "../components/heroes/HeroesScreen/HeroesScreen";
import MarvelScreen from "../components/marvel/MarvelScreen/MarvelScreen";
import SearchScreen from "../components/search/SearchScreen/SearchScreen";
import Navbar from "../components/ui/Navbar";

export const DashboardRoutes = () => {
  return (
    <div className="">
      <Navbar />
      <div className="text-white bg-heroBlack bg-">
        <Switch>
          <Route exact path="/marvel" component={MarvelScreen} />
          <Route exact path="/hero/:heroeId" component={HeroesScreen} />
          <Route exact path="/dc" component={DCScreen} />
          <Route exact path="/search" component={SearchScreen} />
          <Redirect to="/marvel" />
        </Switch>
      </div>
    </div>
  );
};
