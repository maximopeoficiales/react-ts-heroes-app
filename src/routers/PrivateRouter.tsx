import { Redirect, Route, RouteProps } from "react-router-dom";

interface MyProps extends Partial<RouteProps> {
  isAuthenticated: boolean;
  redirectTo: string;
  component: React.ComponentType<any>;
}

export const PrivateRouter = (props: MyProps) => {
  const { isAuthenticated, component: Component, redirectTo, ...rest } = props;

  //   guardo ruta que trata de acceder para futuro uso
  localStorage.setItem("lastPath", rest.location?.pathname ?? "");
  return (
    <Route
      {...rest}
      component={(props: JSX.IntrinsicAttributes) =>
        isAuthenticated ? (
          <Component {...props} />
        ) : (
          <Redirect to={redirectTo} />
        )
      }
    />
  );
};
