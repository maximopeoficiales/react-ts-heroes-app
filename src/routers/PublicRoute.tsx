import { Redirect, Route, RouteProps } from "react-router-dom";

interface MyProps extends Partial<RouteProps> {
  isAuthenticated: boolean;
  redirectTo: string;
  component: React.ComponentType<any>;
}

export const PublicRouter = (props: MyProps) => {
  const { isAuthenticated, component: Component, redirectTo, ...rest } = props;

  return (
    <Route
      {...rest}
      component={(props: JSX.IntrinsicAttributes) =>
        !isAuthenticated ? (
          <Component {...props} />
        ) : (
          <Redirect to={redirectTo} />
        )
      }
    />
  );
};
