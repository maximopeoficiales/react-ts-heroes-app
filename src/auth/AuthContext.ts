import { createContext } from "react";
import { ReducerUser, User } from "./authReducer";

export interface UserProvider {
    user?: User,
    dispatch: React.Dispatch<ReducerUser>
}
export const AuthContext = createContext<UserProvider>({ dispatch: () => { } });