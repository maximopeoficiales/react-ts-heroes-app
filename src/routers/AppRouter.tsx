import React, { useContext } from "react";
import { BrowserRouter as Router, Switch, Route, Link } from "react-router-dom";
import { AuthContext } from "../auth/AuthContext";
import LoginScreen from "../components/login/LoginScreen/LoginScreen";
import Navbar from "../components/ui/Navbar";
import { DashboardRoutes } from "./DashboardRoutes";
import { PrivateRouter } from "./PrivateRouter";
import { PublicRouter } from "./PublicRoute";

export const AppRouter = () => {
  const { user } = useContext(AuthContext);
  return (
    <Router>
      <div className="">
        {/* <Navbar /> */}
        {/* A <Switch> looks through its children <Route>s and
            renders the first one that matches the current URL. */}
        <Switch>
          <PublicRouter
            isAuthenticated={user?.logged ?? false}
            exact
            path="/login"
            component={LoginScreen}
            redirectTo="/"
          />
          <PrivateRouter
            isAuthenticated={user?.logged ?? false}
            path="/"
            component={DashboardRoutes}
            redirectTo={"/login"}
          />
        </Switch>
      </div>
    </Router>
  );
};
