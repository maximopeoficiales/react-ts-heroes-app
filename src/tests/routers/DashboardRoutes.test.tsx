import { mount } from "enzyme";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../auth/AuthContext";
import { DashboardRoutes } from "../../routers/DashboardRoutes";

describe("Pruebas en <DashBoardRoutes/>", () => {
  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: true,
      name: "MaximoProg",
    },
  };

  test("debe mostrarse correctamente el componente", () => {
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter>
          <DashboardRoutes />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").text()).toBe(contextValue.user.name);
  });
});
