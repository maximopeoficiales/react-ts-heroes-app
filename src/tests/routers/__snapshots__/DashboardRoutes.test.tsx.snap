// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DashBoardRoutes/> debe mostrarse correctamente el componente 1`] = `
<div
  className=""
>
  <nav
    className="bg-gray-800"
  >
    <div
      className="px-2 mx-auto max-w-8xl sm:px-6 lg:px-8"
    >
      <div
        className="relative flex items-center justify-between h-16"
      >
        <div
          className="absolute inset-y-0 left-0 flex items-center sm:hidden"
        >
          <button
            aria-controls="mobile-menu"
            aria-expanded="false"
            className="inline-flex items-center justify-center p-2 text-gray-400 rounded-md hover:text-white hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white"
            type="button"
          />
        </div>
        <div
          className="flex items-center justify-center flex-1 sm:items-stretch sm:justify-start"
        >
          <div
            className="flex items-center flex-shrink-0"
          />
          <div
            className="hidden sm:block sm:ml-6"
          >
            <div
              className="flex space-x-4"
            >
              <a
                aria-current="page"
                className="px-3 py-2 text-sm font-medium text-white rounded-md active"
                href="/"
                onClick={[Function]}
                style={Object {}}
              >
                HeroesApp
              </a>
              <a
                aria-current="page"
                className="px-3 py-2 text-sm font-medium text-gray-300 rounded-md hover:bg-gray-700 hover:text-white bg-gray-900"
                href="/marvel"
                onClick={[Function]}
                style={Object {}}
              >
                Marvel
              </a>
              <a
                aria-current={null}
                className="px-3 py-2 text-sm font-medium text-gray-300 rounded-md hover:bg-gray-700 hover:text-white"
                href="/dc"
                onClick={[Function]}
              >
                DC
              </a>
              <a
                aria-current={null}
                className="px-3 py-2 text-sm font-medium text-gray-300 rounded-md hover:bg-gray-700 hover:text-white"
                href="/search"
                onClick={[Function]}
              >
                Search
              </a>
            </div>
          </div>
        </div>
        <div
          className="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
        >
          <p
            className="mx-2 font-bold text-white"
          >
            MaximoProg
          </p>
          <button
            className="px-3 py-2 font-bold text-white duration-75 ease-out delay-75 transform scale-100 bg-green-600 rounded-full active:scale-75 active:bg-green-800 "
            onClick={[Function]}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
    <div
      className="sm:hidden "
      id="mobile-menu"
    >
      <div
        className="px-2 pt-2 pb-3 space-y-1"
      >
        <a
          aria-current="page"
          className="block px-3 py-2 text-base font-medium text-white rounded-md active"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          HeroesApp
        </a>
        <a
          aria-current="page"
          className="block px-3 py-2 text-base font-medium text-gray-300 rounded-md hover:bg-gray-700 hover:text-white bg-gray-900"
          href="/marvel"
          onClick={[Function]}
          style={Object {}}
        >
          Marvel
        </a>
        <a
          aria-current={null}
          className="block px-3 py-2 text-base font-medium text-gray-300 rounded-md hover:bg-gray-700 hover:text-white"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          aria-current={null}
          className="block px-3 py-2 text-base font-medium text-gray-300 rounded-md hover:bg-gray-700 hover:text-white"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
  </nav>
  <div
    className="text-white bg-heroBlack bg-"
  >
    <div
      className=""
      data-testid="MarvelScreen"
    >
      <div
        className="w-11/12 mx-auto my-8 lg:w-9/12"
      >
        <div
          className="p-5 rounded-xl lg:p-14 bg-heroBlackSecondary"
        >
          <div
            className="p-1 my-2 text-4xl font-bold text-center sm:text-left"
            data-testid="Title"
          >
            Marvel
          </div>
          <div
            className="grid grid-cols-1 grid-rows-1 gap-6 md:grid-cols-3 lg:grid-cols-3"
          >
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-spider"
              onClick={[Function]}
            >
              <img
                alt="marvel-spider"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-spider.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Spider Man
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-captain"
              onClick={[Function]}
            >
              <img
                alt="marvel-captain"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-captain.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Captain America
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-iron"
              onClick={[Function]}
            >
              <img
                alt="marvel-iron"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-iron.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Iron Man
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-thor"
              onClick={[Function]}
            >
              <img
                alt="marvel-thor"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-thor.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Thor
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-hulk"
              onClick={[Function]}
            >
              <img
                alt="marvel-hulk"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-hulk.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Hulk
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-wolverine"
              onClick={[Function]}
            >
              <img
                alt="marvel-wolverine"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-wolverine.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Wolverine
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-daredevil"
              onClick={[Function]}
            >
              <img
                alt="marvel-daredevil"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-daredevil.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Daredevil
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-hawkeye"
              onClick={[Function]}
            >
              <img
                alt="marvel-hawkeye"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-hawkeye.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Hawkeye
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-cyclops"
              onClick={[Function]}
            >
              <img
                alt="marvel-cyclops"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-cyclops.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Cyclops
              </div>
            </a>
            <a
              className="inline-flex flex-wrap overflow-hidden bg-gray-100 shadow-lg rounded-xl"
              data-testid="HeroCard"
              href="/hero/marvel-silver"
              onClick={[Function]}
            >
              <img
                alt="marvel-silver"
                className="object-cover w-full rounded-b-none rounded-xl h-96 card-heroe-filter"
                loading="lazy"
                src="/assets/heroes/marvel-silver.jpg"
              />
              <div
                className="block w-full px-3 py-2 text-lg font-bold text-center text-white duration-300 ease-in-out delay-75 transform bg-black"
              >
                Silver Surfer
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
